cyten/tensors.py

```diff
@@ -5087,6 +5087,99 @@ def svd_apply_mask(U: SymmetricTensor, S: DiagonalTensor, Vh: SymmetricTensor, m
     return U, S, Vh
 
 
+def tensor_from_grid(grid: list[list[SymmetricTensor | None]],
+                     labels: Sequence[list[str | None] | None] | list[str | None] | None = None,
+                     dtype: Dtype = Dtype.complex128) -> SymmetricTensor:
+    """Convert a grid of tensors to a single SymmetricTensor.
+
+    The tensors are combined along the first leg in their codomain and the final leg in their
+    domain. All other legs must be identical across all tensors in the grid. For consistency,
+    tensors within the same row must have identical left spaces (first leg in the codomain),
+    and tensors within the same column must have identical right spaces (final leg in the
+    domain).
+
+    Parameters
+    ----------
+    grid: list[list[SymmetricTensor | None]]
+        Contains the tensors from which a single tensor is constructed. `None` entries are
+        interpreted as tensors with all blocks equal to zero.
+    labels
+        Leg labels of the resulting tensor.
+    dtype: Dtype
+        The dtype of the tensor.
+    """
+    op_list = [op for row in grid for op in row if op is not None]
+    backend = get_same_backend(*op_list)
+    device = get_same_device(*op_list)
+    # check input
+    for op in op_list:
+        assert op.num_codomain_legs == op_list[0].num_codomain_legs
+        assert op.num_domain_legs == op_list[0].num_domain_legs
+        assert op.codomain[1:] == op_list[0].codomain[1:]
+        assert op.domain[:-1] == op_list[0].domain[:-1]
+        # only ElementarySpaces have direct_sum
+        assert isinstance(op.codomain[0], ElementarySpace)
+        assert isinstance(op.domain[-1], ElementarySpace)
+
+    transposed_grid = list(map(list, zip(*grid)))
+    right_ops = grid[0]
+    for i, op in enumerate(right_ops):
+        if op is not None:
+            continue
+        # find op from same column
+        for new_op in transposed_grid[i]:
+            if new_op is None:
+                continue
+            right_ops[i] = new_op
+            break
+    right_spaces = [op.domain[-1] for op in right_ops]
+    if None in right_spaces:
+        raise ValueError('One column in the grid does not have nonzero entries.')
```

Comment:
Formulate positively: "Must have at least one nonzero entry in each column"
same below for rows